### ISSUE TITLE
Query: client eval when trying to bind to a property from subquery that is part of a JOIN

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#10812]
Repro:

```cs
from ll in ctx.LocustLeaders
join h in ctx.Factions.OfType<LocustHorde>().Select(f => new { IsEradicated = f.Name == "Locust" ? (bool?)true : null, f.CommanderName, f.Name }) on ll.Name equals h.CommanderName
where h.IsEradicated != true
select h;
```

query plan:

```
(QueryContext queryContext) => IEnumerable<<>f__AnonymousType255<Nullable<bool>, string, string>> _InterceptExceptions(
    source: IEnumerable<<>f__AnonymousType255<Nullable<bool>, string, string>> _Select(
        source: IEnumerable<TransparentIdentifier<ValueBuffer, <>f__AnonymousType255<Nullable<bool>, string, string>>> _Where(
            source: IEnumerable<TransparentIdentifier<ValueBuffer, <>f__AnonymousType255<Nullable<bool>, string, string>>> _ShapedQuery(
                queryContext: queryContext, 
                shaperCommandContext: SelectExpression: 
                    SELECT [t].[IsEradicated], [t].[CommanderName], [t].[Name]
                    FROM [LocustLeaders] AS [ll]
                    INNER JOIN (
                        SELECT CASE
                            WHEN [f].[Name] = N'Locust'
                            THEN CAST(1 AS BIT) ELSE NULL
                        END AS [IsEradicated], [f].[CommanderName], [f].[Name]
                        FROM [Factions] AS [f]
                        WHERE [f].[Discriminator] = N'LocustHorde'
                    ) AS [t] ON [ll].[Name] = [t].[CommanderName]
                    WHERE [ll].[Discriminator] IN (N'LocustCommander', N'LocustLeader'), 
                shaper: TypedCompositeShaper<ValueBufferShaper, ValueBuffer, TypedProjectionShaper<ValueBufferShaper, ValueBuffer, <>f__AnonymousType255<Nullable<bool>, string, string>>, <>f__AnonymousType255<Nullable<bool>, string, string>, TransparentIdentifier<ValueBuffer, <>f__AnonymousType255<Nullable<bool>, string, string>>>), 
            predicate: (TransparentIdentifier<ValueBuffer, <>f__AnonymousType255<Nullable<bool>, string, string>> t0) => t0.Inner.IsEradicated != True), 
        selector: (TransparentIdentifier<ValueBuffer, <>f__AnonymousType255<Nullable<bool>, string, string>> t0) => t0.Inner), 
    contextType: TestModels.GearsOfWarModel.GearsOfWarContext, 
    logger: DiagnosticsLogger<Query>, 
    queryContext: queryContext)
```

Filter after the join is client eval-ed because we cant find mapping to IsEradicated - it is lost as some point during join processing.

**Comment by smitpatel** [comment]
related #8094

**Comment by ajcvickers** [comment]
Note from triage: this issue may be fixed when #8094 is fixed, but this should be tested, so leaving this issue open as a bug.

**Comment by smitpatel** [comment]
Generated SQL
```
            AssertSql(
                @"SELECT CASE
    WHEN [t].[Name] = N'Locust' THEN CAST(1 AS bit)
    ELSE NULL
END AS [IsEradicated], [t].[CommanderName], [t].[Name]
FROM [LocustLeaders] AS [l]
INNER JOIN (
    SELECT [f].[Id], [f].[CapitalName], [f].[Discriminator], [f].[Name], [f].[CommanderName], [f].[Eradicated]
    FROM [Factions] AS [f]
    WHERE ([f].[Discriminator] = N'LocustHorde') AND ([f].[Discriminator] = N'LocustHorde')
) AS [t] ON [l].[Name] = [t].[CommanderName]
WHERE [l].[Discriminator] IN (N'LocustLeader', N'LocustCommander') AND ((CASE
    WHEN [t].[Name] = N'Locust' THEN CAST(1 AS bit)
    ELSE NULL
END <> CAST(1 AS bit)) OR CASE
    WHEN [t].[Name] = N'Locust' THEN CAST(1 AS bit)
    ELSE NULL
END IS NULL)");
```
